vnext/Microsoft.ReactNative/JsiWriter.cpp

```diff
@@ -0,0 +1,146 @@
+}
+
+void JsiWriter::WriteBoolean(bool value) noexcept {
+  WriteValue({value});
+}
+
+void JsiWriter::WriteInt64(int64_t value) noexcept {
+  // JavaScript's integer is not int64_t, need to ensure that the value is in range
+  VerifyElseCrash(MinSafeInteger <= value && value <= MaxSafeInteger);
+
+  // unfortunately JSI only supports int and double for number, choose double here
+  // make sure that doing type conversion in C++ makes no data loss
+  double d = static_cast<double>(value);
+  VerifyElseCrash(floor(d) == d);
+  WriteValue(d);
+}
+
+void JsiWriter::WriteDouble(double value) noexcept {
+  WriteValue({value});
+}
+
+void JsiWriter::WriteString(const winrt::hstring &value) noexcept {
+  WriteValue({m_runtime, facebook::jsi::String::createFromUtf8(m_runtime, winrt::to_string(value))});
+}
+
+void JsiWriter::WriteObjectBegin() noexcept {
+  // legal to create an object when it is accepting a value
+  VerifyElseCrash(Top().Action != ContinuationAction::AcceptPropertyName);
+  Push({ContinuationAction::AcceptPropertyName, {m_runtime, facebook::jsi::Object(m_runtime)}});
+}
+
+void JsiWriter::WritePropertyName(const winrt::hstring &name) noexcept {
+  // legal to set a property name only when AcceptPropertyName
+  auto &top = Top();
+  VerifyElseCrash(top.Action == ContinuationAction::AcceptPropertyName);
+  top.Action = ContinuationAction::AcceptPropertyValue;
+  top.PropertyName = winrt::to_string(name);
+}
+
+void JsiWriter::WriteObjectEnd() noexcept {
+  // legal to finish an object only when AcceptPropertyName
+  VerifyElseCrash(Top().Action == ContinuationAction::AcceptPropertyName);
+  auto createdObject = std::move(Pop().Values.at(0));
+  WriteValue(std::move(createdObject));
+}
+
+void JsiWriter::WriteArrayBegin() noexcept {
+  // legal to create an array only when it is accepting a value
+  VerifyElseCrash(Top().Action != ContinuationAction::AcceptPropertyName);
+  Push({ContinuationAction::AcceptArrayElement});
+}
+
+void JsiWriter::WriteArrayEnd() noexcept {
+  // legal to finish an array only when AcceptArrayElement
+  auto &top = Top();
+  VerifyElseCrash(top.Action == ContinuationAction::AcceptArrayElement);
+  facebook::jsi::Array createdArray(m_runtime, top.Values.size());
+  for (size_t i = 0; i < top.Values.size(); i++) {
+    createdArray.setValueAtIndex(m_runtime, i, std::move(top.Values.at(i)));
+  }
+  Pop();
+  WriteValue({m_runtime, createdArray});
+}
+
+void JsiWriter::WriteValue(facebook::jsi::Value &&value) noexcept {
+  auto &top = Top();
+  switch (top.Action) {
+    case ContinuationAction::AcceptValueAndFinish: {
+      m_result = std::move(value);
+      Pop();
+      VerifyElseCrash(m_continuations.size() == 0);
+      break;
+    }
+    case ContinuationAction::AcceptArrayElement: {
+      top.Values.push_back(std::move(value));
+      break;
+    }
+    case ContinuationAction::AcceptPropertyValue: {
+      auto createdObject = top.Values.at(0).getObject(m_runtime);
```

Comment:
>top.Values.at(0) [](start = 27, length = 16)

Using vector entry for an object causes unnecessary memory allocation.

Reply:
I am also considering using optional<Object>+optional<vector<Value>> to replace this vector. What do you think?

Reply:
The optional<Object> makes sense. The vector does not need to be optional. Empty vector does not allocate in heap.

---
In reply to: [430724697](https://github.com/microsoft/react-native-windows/pull/5009#discussion_r430724697) [](ancestors = 430724697)